### PR TITLE
Add complete CI

### PR DESCRIPTION
**Pull request by szepeviktor**
@tillkruss Please add this repo to your Coveralls account.
https://coveralls.io/repos/new

### CODE REVIEW

**Comment by szepeviktor** [comment]
@tillkruss This PR will be so beautiful.

**Comment by szepeviktor** [comment]
@tillkruss Would you like me to move TODO-s (from at least src/) to GitHub issues?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ on:
     branches:
       - main
       - v2.**
-  pull_request:
+#  pull_request:
```
Comment (szepeviktor) [inline .github/workflows/tests.yml:8]: This will be reverted ⚠️

File: composer.json [hunk 2]
```diff
@@ -26,6 +26,7 @@
     },
     "require-dev": {
         "friendsofphp/php-cs-fixer": "^3.3",
+        "php-parallel-lint/php-parallel-lint": "^1.3",
```
Comment (tillkruss) [inline composer.json:29]: What do we need this for?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (tillkruss) [inline .github/workflows/tests.yml:8]: Why the removal?

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -59,9 +59,15 @@ jobs:
...
       - name: Install Composer dependencies
-        env:
-          PHP_VERSION: ${{ matrix.php }}
-        run: composer install --ansi --no-progress --prefer-dist $(if [ "$PHP_VERSION" == "8.0" ]; then echo "--ignore-platform-reqs"; fi;)
+        run: composer install --ansi --no-progress --prefer-dist ${{ matrix.php == '8.0' && '--ignore-platform-reqs' || '' }}
 
       - name: Run PHPUnit tests
         run: vendor/bin/phpunit
+
+      - name: Send coverage to Coveralls
```
Comment (tillkruss) [inline .github/workflows/tests.yml:67]: What's this service?

File: .github/workflows/linters.yml [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      - "main"
+      - "v2.**"
+  pull_request: null
+# Add [skip ci] to commit message to skip CI.
+
+permissions: {}
+
+concurrency:
+  group: "${{ github.workflow }}-${{ github.ref }}"
+  cancel-in-progress: true
```
Comment (tillkruss) [inline .github/workflows/linters.yml:17]: I think the other workflows need this too.

File: .github/workflows/linters.yml [hunk 5]
```diff
@@ -0,0 +1,161 @@
+# yaml-language-server: $schema=https://json.schemastore.org/github-workflow
```
Comment (tillkruss) [inline .github/workflows/linters.yml:1]: Can not not enforce the double quotes in YAML, I dislike the visual clutter.

File: .github/workflows/linters.yml [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    #strategy:
+    #  #fail-fast: false
+    #  matrix:
+    #    php-version:
+    #      - "8.0"
+    #    dependencies:
+    #      - "lowest"
+    #      - "highest"
+    #runs-on: "ubuntu-latest"
+    #steps:
```
Comment (szepeviktor) [inline .github/workflows/linters.yml:67]: May I include unit tests in this workflow?
That would stop CI when there is a 0️⃣ Byte-level or 1️⃣ Syntax error.

File: .editorconfig [hunk 7]
```diff
@@ -9,9 +9,12 @@ indent_size = 4
 end_of_line = lf
 insert_final_newline = true
 trim_trailing_whitespace = true
+block_comment_start = /*
+block_comment = *
+block_comment_end = */
 
 [*.php]
-max_line_length = 150
+max_line_length = 223
```
Comment (tillkruss) [inline .editorconfig:17]: Is this a hot fix?

File: composer.json (hunk 2 above)
Comment (szepeviktor) [reply composer.json:29]: For the "1️⃣ Syntax error" job.

Comment (szepeviktor) [reply composer.json:29]: kép

File: .github/workflows/tests.yml (hunk 1 above)
Comment (szepeviktor) [reply .github/workflows/tests.yml:8]: This will be of course reverted.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (szepeviktor) [reply .github/workflows/tests.yml:67]: Coveralls shows historical and per file coverage.
https://github.com/predis/predis/actions/runs/3972188215/jobs/6809833063

File: .github/workflows/linters.yml (hunk 4 above)
Comment (szepeviktor) [reply .github/workflows/linters.yml:17]: Done.

File: .editorconfig (hunk 7 above)
Comment (szepeviktor) [reply .editorconfig:17]: Yes. In docblocks the whole function signature must be on 1 line.

File: .github/workflows/linters.yml (hunk 5 above)
Comment (szepeviktor) [reply .github/workflows/linters.yml:1]: This makes Viktor very sad.

Arrives in a minute.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (szepeviktor) [reply .github/workflows/tests.yml]: Reverted.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (szepeviktor) [reply .github/workflows/tests.yml:78]: Please see what a beauty she is https://coveralls.io/jobs/113336063
You can open the directory tree and see where predis may improve.

**Comment by tillkruss** [comment]
> @tillkruss Would you like me to move TODO-s (from at least src/) to GitHub issues?

Yes, let's do that.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (tillkruss) [reply .github/workflows/tests.yml:69]: Can we get the source code available?

https://coveralls.io/jobs/113336063/source_files/2021808842

File: .editorconfig (hunk 7 above)
Comment (tillkruss) [reply .editorconfig:17]: Let's keep it at 150?
But set a custom limit for that one file to 1000?

**Comment by szepeviktor** [comment]
> Yes, let's do that.

Done 🟢

File: .editorconfig (hunk 7 above)
Comment (szepeviktor) [reply .editorconfig:17]: I think https://github.com/greut/eclint does not have that feature.

Getting it done in a minute ... 🏃🏻‍♂️

File: .github/workflows/tests.yml (hunk 3 above)
Comment (szepeviktor) [reply .github/workflows/tests.yml:69]: Yes. Please add this repo to your Coveralls account.
https://github.com/predis/predis/pull/1036?notification_referrer_id=NT_kwDOAA6Gx7E1Mzg1OTg5NzYyOjk1MjAwNw#issue-1551602274

File: composer.json (hunk 2 above)
Comment (tillkruss) [reply composer.json:29]: Can we use setup-php's `tools` for `parallel-lint`, since we don't run this locally using composer?

Comment (szepeviktor) [reply composer.json:29]: That goes against viktor dependency policy.

Will be ready in minutes ... 🏃🏻‍♂️

File: .github/workflows/linters.yml (hunk 6 above)
Comment (szepeviktor) [reply .github/workflows/linters.yml]: @tillkruss How about this one? Merging two workflows for the above reason?

File: .github/workflows/tests.yml (hunk 3 above)
Comment (szepeviktor) [reply .github/workflows/tests.yml:69]: We could have a badge too.
```
[![Coverage Status](https://coveralls.io/repos/github/predis/predis/badge.svg?branch=main)](https://coveralls.io/github/predis/predis?branch=main)
```
[Coverage Status](https://coveralls.io/github/predis/predis?branch=main)

File: .editorconfig [hunk 8]
```diff
@@ -19,16 +17,8 @@ max_line_length = 150
...
-[tests/Predis/Configuration/Option/AggregateTest.php]
+[tests/**.php]
 max_line_length = unset
-[tests/Predis/Configuration/Option/ClusterTest.php]
-max_line_length = unset
-[src/ClientContextInterface.php]
-max_line_length = unset
-[src/ClientInterface.php]
+
+[{src/ClientInterface.php, src/ClientContextInterface.php}]
```
Comment (szepeviktor) [inline .editorconfig:23]: Maybe that space makes eclint fail.

File: .editorconfig [hunk 9]
```diff
@@ -9,9 +7,18 @@ indent_size = 4
...
 [*.php]
 max_line_length = 150
 
 [*.{md,yml,yaml,neon}]
 indent_size = 2
+
+[tests/**.php]
+max_line_length = unset
+
+[src/ClientInterface.php,src/ClientContextInterface.php]
```
Comment (szepeviktor) [inline .editorconfig:23]:
```suggestion
[{src/ClientInterface.php,src/ClientContextInterface.php}]
```

✈️

File: .github/workflows/linters.yml (hunk 6 above)
Comment (tillkruss) [reply .github/workflows/linters.yml]: No, let's keep those separate.

**Review by tillkruss — APPROVED** [review]
This is amazing, as always.

File: .github/workflows/linters.yml (hunk 6 above)
Comment (szepeviktor) [reply .github/workflows/linters.yml]: 🖖🏻

**Comment by szepeviktor** [comment]
It feels good to work for you.

**Pull request merged**